### ISSUE TITLE
Ignore type-checking only imports in cycle detection

### ISSUE THREAD

**Issue by mariusmuja** [microsoft/pyright#1825]
**Is your feature request related to a problem? Please describe.**

Sometimes for type-checking purposes, it's needed/easier to use circular imports. To avoid run-time errors, I'm using the `TYPE_CHECKING` constant to break import cycles at run-time and only have them during type checking (as also suggested in PEP 484). These cycle are however still reported by pyright.

Code example:
```python
# file a.py
from .b import B

class A(object):

  def __init__(self, b: B) -> None:
    b.parent = self
    self.b = b
```
```python
# file b.py
from typing import TYPE_CHECKING, Union
if TYPE_CHECKING:
  from .a import A

class B(object):

  def __init__(self) -> None:
    self.parent: Union[A, None] = None
``` 

**Describe the solution you'd like**
It would be preferable that pyright ignores imports inside an `if TYPE_CHECKING:` block.

**Comment by erictraut** [comment]
When you say "still reported by pyright", I presume that you are referring to the `reportImportCycles` diagnostic check. Is that correct?

The `reportImportCycles` check in pyright was designed to detect architectural layering violations that manifest in the form of cyclical dependencies. You have such a layering violation in your code, and you're trying to work around it using a `TYPE_CHECKING` conditional. If you are interested in architectural layering, you should refactor your code to avoid the layering violation. If you don't care about architectural layering, then you should disable the `reportImportCycles` diagnostic rule.

I'll also point out that it would be a dangerous precedent for the type checker to honor code within `TYPE_CHECKING` blocks in some cases but ignore it in others. I don't think that's a good idea. It would go against the intent of the `TYPE_CHECKING` symbol as defined in PEP 484.

**Comment by mariusmuja** [comment]
> When you say "still reported by pyright", I presume that you are referring to the reportImportCycles diagnostic check. Is that correct?

Yes, exactly. 

> The reportImportCycles check in pyright was designed to detect architectural layering violations that manifest in the form of cyclical dependencies. You have such a layering violation in your code, and you're trying to work around it using a TYPE_CHECKING conditional. If you are interested in architectural layering, you should refactor your code to avoid the layering violation. If you don't care about architectural layering, then you should disable the reportImportCycles diagnostic rule.

I agree, refactoring the code is preferable. I'm in the process of adding type checking annotation to an existing code base and I wanted to have the "reportImportCycles" diagnostic enabled to make sure I don't add circular imports by mistake while doing this.

> I'll also point out that it would be a dangerous precedent for the type checker to honor code within TYPE_CHECKING blocks in some cases but ignore it in others. I don't think that's a good idea. It would go against the intent of the TYPE_CHECKING symbol as defined in PEP 484.

Yes, this makes perfect sense if "reportImportCycles" is only designed to detect architectural layering violations. There is also the use case when architectural layering violations exist in the code base and detecting run-time circular imports would be very useful.

Thanks for the explanation.

**Comment by coder0xff** [comment]
There's a middle ground. If reportImportCycles must insist on telling users their architectural layering should be refactored, consider a different setting to honor TYPE_CHECKING. There are cases where this pattern is useful (such as return types), and arguing over whether nuances exist is unproductive. Let's look at solutions everyone can be happy with.
